File: HandSock-Server/src/main/java/com/server/handsock/console/IDGenerator.java
package com.server.handsock.console;

import org.springframework.stereotype.Service;

import java.util.Calendar;
import java.util.Random;
import java.util.UUID;

@Service
public class IDGenerator {
    private final UtilityService utilityService = new UtilityService();

    public long generateRandomId(int length) {
        int currentYear = Calendar.getInstance().get(Calendar.YEAR);
        String yearPrefix = String.valueOf(currentYear).substring(currentYear - 2000 > 0 ? 2 : 0);
        Random random = new Random();
        String randomSuffix = String.format("%0" + length + "d", random.nextInt(1000000));
        return Long.parseLong(yearPrefix + randomSuffix);
    }

    public String generateRandomMessageId(Long uid, long gid, String address) {
        return utilityService.encodeStringToMD5(UUID.randomUUID() + "-" + utilityService.encodeStringToMD5(uid + String.valueOf(gid) + address));
    }

    public String generateRandomReportedId(String sid, Long reporter_id, Long reported_id) {
        return utilityService.encodeStringToMD5(UUID.randomUUID() + "-" + utilityService.encodeStringToMD5(sid + reporter_id + reported_id));
    }

    public String generateRandomFileId(Long uid, String name, String path, String time) {
        return utilityService.encodeStringToMD5(UUID.randomUUID() + "-" + utilityService.encodeStringToMD5(uid + name + path + time));
    }
}

File: HandSock-Server/src/main/java/com/server/handsock/utils/IDGenerator.kt
package com.server.handsock.utils

import java.util.*

object IDGenerator {
    fun generateUniqueId(): String {
        return UUID.randomUUID().toString()
    }

    fun generateRandomId(length: Int): Long {
        val currentYear = Calendar.getInstance()[Calendar.YEAR]
        val yearPrefix = currentYear.toString().substring(if (currentYear - 2000 > 0) 2 else 0)
        val random = Random()
        val randomSuffix = String.format("%0" + length + "d", random.nextInt(1000000))
        return (yearPrefix + randomSuffix).toLong()
    }

    fun generateRandomMessageId(uid: Long, gid: Long, address: String): String {
        return HandUtils.encodeStringToMD5(
            UUID.randomUUID().toString() + "-" + HandUtils.encodeStringToMD5(
                uid.toString() + gid.toString() + address
            )
        )
    }

    fun generateRandomReportedId(sid: String, reporterId: Long, reportedId: Long): String {
        return HandUtils.encodeStringToMD5(
            UUID.randomUUID().toString() + "-" + HandUtils.encodeStringToMD5(
                sid + reporterId + reportedId
            )
        )
    }

    fun generateRandomFileId(uid: Long, name: String, path: String, time: String): String {
        return HandUtils.encodeStringToMD5(
            UUID.randomUUID().toString() + "-" + HandUtils.encodeStringToMD5(
                uid.toString() + name + path + time
            )
        )
    }
}